File: mobile-rn/src/screens/LoginScreen.js
import React, { useState } from "react";
import {
  View, Text, TextInput, TouchableOpacity, ScrollView, ActivityIndicator, StyleSheet,
} from "react-native";

import { apiPost } from "../api";
import { sessionFromAuth } from "../session";
import { colors } from "../theme";

export default function LoginScreen({ onLogin }) {
  const [signup, setSignup] = useState(false);
  const [role, setRole] = useState("patient");
  const [busy, setBusy] = useState(false);
  const [error, setError] = useState(null);

  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [age, setAge] = useState("");
  const [gender, setGender] = useState("");
  const [specialization, setSpecialization] = useState("");

  const submit = async () => {
    setBusy(true); setError(null);
    try {
      let data;
      if (signup) {
        const body = {
          username: username.trim(), password, role,
          name: name.trim() || null, email: email.trim() || null,
        };
        if (role === "patient") {
          body.age = age.trim() ? Number(age.trim()) : null;
          body.gender = gender || null;
        } else {
          body.specialization = specialization.trim() || null;
        }
        data = await apiPost("/auth/register", body);
      } else {
        data = await apiPost("/auth/login", { username: username.trim(), password });
      }
      await onLogin(sessionFromAuth(data));
    } catch (e) {
      setError(e.message || String(e));
    } finally {
      setBusy(false);
    }
  };

  const Seg = ({ value, current, onPress, label }) => (
    <TouchableOpacity
      style={[styles.seg, current === value && styles.segActive]}
      onPress={() => onPress(value)}
    >
      <Text style={[styles.segText, current === value && styles.segTextActive]}>{label}</Text>
    </TouchableOpacity>
  );

  return (
    <ScrollView contentContainerStyle={styles.wrap}>
      <View style={styles.card}>
        <Text style={styles.brand}>DermAI</Text>
        <Text style={styles.sub}>AI-powered skin disease screening</Text>

        <View style={styles.segRow}>
          <Seg value={false} current={signup} onPress={setSignup} label="Sign in" />
          <Seg value={true} current={signup} onPress={setSignup} label="Create account" />
        </View>

        {signup && (
          <View style={styles.segRow}>
            <Seg value="patient" current={role} onPress={setRole} label="🧑 Patient" />
            <Seg value="doctor" current={role} onPress={setRole} label="👨‍⚕️ Doctor" />
          </View>
        )}

        <Text style={styles.label}>Username</Text>
        <TextInput style={styles.input} autoCapitalize="none" value={username} onChangeText={setUsername} />
        <Text style={styles.label}>Password</Text>
        <TextInput style={styles.input} secureTextEntry value={password} onChangeText={setPassword} />

        {signup && (
          <>
            <Text style={styles.label}>Full name</Text>
            <TextInput style={styles.input} value={name} onChangeText={setName} />
            <Text style={styles.label}>Email</Text>
            <TextInput style={styles.input} autoCapitalize="none" keyboardType="email-address" value={email} onChangeText={setEmail} />
            {role === "patient" ? (
              <>
                <Text style={styles.label}>Age</Text>
                <TextInput style={styles.input} keyboardType="number-pad" value={age} onChangeText={setAge} />
                <Text style={styles.label}>Gender</Text>
                <View style={styles.segRow}>
                  {["Male", "Female", "Other"].map((g) => (
                    <Seg key={g} value={g} current={gender} onPress={setGender} label={g} />
                  ))}
                </View>
              </>
            ) : (
              <>
                <Text style={styles.label}>Specialization</Text>
                <TextInput style={styles.input} value={specialization} onChangeText={setSpecialization} />
                <Text style={styles.note}>Doctor accounts require admin verification.</Text>
              </>
            )}
          </>
        )}

        {error && <Text style={styles.error}>{error}</Text>}

        <TouchableOpacity style={styles.button} onPress={submit} disabled={busy}>
          {busy ? <ActivityIndicator color="#fff" />
                : <Text style={styles.buttonText}>{signup ? "Create account" : "Sign in"}</Text>}
        </TouchableOpacity>
      </View>
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  wrap: { flexGrow: 1, justifyContent: "center", padding: 20, backgroundColor: colors.bg },
  card: { backgroundColor: colors.card, borderRadius: 20, padding: 22, borderWidth: 1, borderColor: colors.line },
  brand: { fontSize: 28, fontWeight: "800", color: colors.patient, textAlign: "center" },
  sub: { color: colors.muted, textAlign: "center", marginTop: 2, marginBottom: 16 },
  segRow: { flexDirection: "row", backgroundColor: "#F1F5F9", borderRadius: 12, padding: 4, marginBottom: 10 },
  seg: { flex: 1, paddingVertical: 9, borderRadius: 9, alignItems: "center" },
  segActive: { backgroundColor: "#fff" },
  segText: { color: colors.muted, fontWeight: "600", fontSize: 12 },
  segTextActive: { color: colors.ink },
  label: { fontSize: 12, fontWeight: "600", color: colors.muted, marginTop: 10, marginBottom: 4 },
  input: { borderWidth: 1, borderColor: colors.line, borderRadius: 10, paddingHorizontal: 12, paddingVertical: 10, backgroundColor: "#fff" },
  note: { fontSize: 12, color: colors.muted, marginTop: 8 },
  error: { color: colors.err, marginTop: 12 },
  button: { backgroundColor: colors.patient, borderRadius: 12, paddingVertical: 14, alignItems: "center", marginTop: 18 },
  buttonText: { color: "#fff", fontWeight: "700", fontSize: 15 },
});

File: mobile/lib/screens/login_screen.dart
import 'package:flutter/material.dart';

import '../services/api_client.dart';
import '../services/session.dart';
import '../theme.dart';

class LoginScreen extends StatefulWidget {
  final Future<void> Function(Session) onLogin;
  const LoginScreen({super.key, required this.onLogin});

  @override
  State<LoginScreen> createState() => _LoginScreenState();
}

class _LoginScreenState extends State<LoginScreen> {
  bool _signup = false;
  String _role = 'patient';
  bool _busy = false;
  String? _error;

  final _username = TextEditingController();
  final _password = TextEditingController();
  final _name = TextEditingController();
  final _email = TextEditingController();
  final _age = TextEditingController();
  final _specialization = TextEditingController();
  String? _gender;

  @override
  void dispose() {
    for (final c in [_username, _password, _name, _email, _age, _specialization]) {
      c.dispose();
    }
    super.dispose();
  }

  Future<void> _submit() async {
    setState(() {
      _busy = true;
      _error = null;
    });
    try {
      final Map<String, dynamic> data;
      if (_signup) {
        final body = <String, dynamic>{
          'username': _username.text.trim(),
          'password': _password.text,
          'role': _role,
          'name': _name.text.trim().isEmpty ? null : _name.text.trim(),
          'email': _email.text.trim().isEmpty ? null : _email.text.trim(),
        };
        if (_role == 'patient') {
          body['age'] = _age.text.trim().isEmpty ? null : int.tryParse(_age.text.trim());
          body['gender'] = _gender;
        } else {
          body['specialization'] =
              _specialization.text.trim().isEmpty ? null : _specialization.text.trim();
        }
        data = await ApiClient.post('/auth/register', body) as Map<String, dynamic>;
      } else {
        data = await ApiClient.post('/auth/login', {
          'username': _username.text.trim(),
          'password': _password.text,
        }) as Map<String, dynamic>;
      }
      await widget.onLogin(Session.fromAuth(data));
    } catch (e) {
      setState(() => _error = e.toString());
    } finally {
      if (mounted) setState(() => _busy = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(20),
          child: ConstrainedBox(
            constraints: const BoxConstraints(maxWidth: 420),
            child: Card(
              child: Padding(
                padding: const EdgeInsets.all(22),
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  children: [
                    const Center(
                      child: Text('DermAI',
                          style: TextStyle(fontSize: 28, fontWeight: FontWeight.w800, color: kPatient)),
                    ),
                    const SizedBox(height: 4),
                    const Center(
                        child: Text('AI-powered skin disease screening',
                            style: TextStyle(color: Colors.black54))),
                    const SizedBox(height: 18),

                    // Sign in / Create account toggle
                    SegmentedButton<bool>(
                      segments: const [
                        ButtonSegment(value: false, label: Text('Sign in')),
                        ButtonSegment(value: true, label: Text('Create account')),
                      ],
                      selected: {_signup},
                      onSelectionChanged: (s) => setState(() => _signup = s.first),
                    ),
                    const SizedBox(height: 14),

                    if (_signup) ...[
                      SegmentedButton<String>(
                        segments: const [
                          ButtonSegment(value: 'patient', label: Text('🧑 Patient')),
                          ButtonSegment(value: 'doctor', label: Text('👨‍⚕️ Doctor')),
                        ],
                        selected: {_role},
                        onSelectionChanged: (s) => setState(() => _role = s.first),
                      ),
                      const SizedBox(height: 12),
                    ],

                    TextField(
                      controller: _username,
                      decoration: const InputDecoration(labelText: 'Username'),
                    ),
                    const SizedBox(height: 10),
                    TextField(
                      controller: _password,
                      obscureText: true,
                      decoration: const InputDecoration(labelText: 'Password'),
                    ),

                    if (_signup) ...[
                      const SizedBox(height: 10),
                      TextField(controller: _name, decoration: const InputDecoration(labelText: 'Full name')),
                      const SizedBox(height: 10),
                      TextField(
                          controller: _email,
                          keyboardType: TextInputType.emailAddress,
                          decoration: const InputDecoration(labelText: 'Email')),
                      if (_role == 'patient') ...[
                        const SizedBox(height: 10),
                        Row(children: [
                          Expanded(
                            child: TextField(
                                controller: _age,
                                keyboardType: TextInputType.number,
                                decoration: const InputDecoration(labelText: 'Age')),
                          ),
                          const SizedBox(width: 10),
                          Expanded(
                            child: DropdownButtonFormField<String>(
                              value: _gender,
                              decoration: const InputDecoration(labelText: 'Gender'),
                              items: const [
                                DropdownMenuItem(value: 'Male', child: Text('Male')),
                                DropdownMenuItem(value: 'Female', child: Text('Female')),
                                DropdownMenuItem(value: 'Other', child: Text('Other')),
                              ],
                              onChanged: (v) => setState(() => _gender = v),
                            ),
                          ),
                        ]),
                      ] else ...[
                        const SizedBox(height: 10),
                        TextField(
                            controller: _specialization,
                            decoration: const InputDecoration(labelText: 'Specialization')),
                        const Padding(
                          padding: EdgeInsets.only(top: 8),
                          child: Text('Doctor accounts require admin verification.',
                              style: TextStyle(fontSize: 12, color: Colors.black54)),
                        ),
                      ],
                    ],

                    if (_error != null) ...[
                      const SizedBox(height: 12),
                      Text(_error!, style: const TextStyle(color: kErr, fontSize: 13)),
                    ],
                    const SizedBox(height: 18),
                    FilledButton(
                      onPressed: _busy ? null : _submit,
                      child: _busy
                          ? const SizedBox(
                              height: 18, width: 18, child: CircularProgressIndicator(strokeWidth: 2, color: Colors.white))
                          : Text(_signup ? 'Create account' : 'Sign in'),
                    ),
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
